File: array/intersect.js
"use strict";

/**accepts: two integer arrays nums1 and nums2
 * returns: an array of their intersection. 
 * Each element in the result must appear as many times as it shows in both arrays and you may return the result in any order.
 * 
 * test case: 
 * nums1 = [1,2,2,1], nums2 = [2,2] => [2, 2]
 * nums1 = [4,9,5], nums2 = [9,4,9,8,4] => [4, 9]
 * 
 * pseudocode:
 * sort two arrays, nums1, nums2
 * make a variable result = []
 * loop through two arrays from the start, make sure the index is less than 
 * corresponding array length
 * if nums1[i] === nums2[j], push the number to result, both index increase 1
 * if less: i++
 * if great: j++
 * 
 * return result
 */

// runtime: O(nlogn) | space complexity: o(n), n is the length of the shorter array
function intersect(nums1, nums2) {
  //make sure add the arrow function inside of sort(), if doesn't add it, 
  //sort function will convert number to string and then compare
  nums1.sort((a, b) => a - b);
  nums2.sort((a, b) => a - b);

  let result = [];
  let i = 0; //track index of nums1
  let j = 0; //track index of nums2

  while (i < nums1.length && j < nums2.length) {
    if (nums1[i] === nums2[j]) {
      result.push(nums1[i]);
      i++;
      j++;
    } else if (nums1[i] < nums2[j]) {
      i++;
    } else {
      j++;
    }
  }
  return result;
}

File: array/intersect.py
from typing import List, Dict

# runtime: O(nlogn) | space: o(n)
# solution1: sort the array
def intersect(nums1: List[int], nums2: List[int]) -> List[int]:
    result: List[int] = []
    nums1.sort()
    nums2.sort()
    
    i: int = 0
    j: int = 0
    
    while i < len(nums1) and j < len(nums2):
        if (nums1[i] == nums2[j]):
            result.append(nums1[i])
            i += 1
            j += 1
        elif (nums1[i] < nums2[j]):
            i += 1
        else:
            j += 1
    
    return result


# runtime: O(n) | space: o(n) + o(m)
# solution1: sort the array
def intersect2(nums1: List[int], nums2: List[int]) -> List[int]: 
    freq1: Dict[int, int] = dict()
    result: List[int] = list()

    for num in nums1:
        freq1[num] = freq1.get(num, 0) + 1

    for num in nums2:
        if num in freq1 and freq1[num] > 0:
            result.append(num)
            freq1[num] -= 1
    return result